### PR TITLE
Support for windows

### PR DESCRIPTION
**Pull request by roznawsk**
Fixed Beamchmark crashing on Windows.
Fixed Beamchmark not opening HTML report due to wrong path format.
Added average CPU utilization measurement for Windows.

### CODE REVIEW

File: lib/beamchmark/suite/cpu/cpu_task.ex [hunk 1]
```diff
@@ -56,12 +57,27 @@ defmodule Beamchmark.Suite.CPU.CpuTask do
...
   defp cpu_snapshot() do
-    to_cpu_snapshot(:cpu_sup.util([:per_cpu]))
+    cpu_snapshot(Utils.os())
   end
 
-  # Converts output of `:cpu_sup.util([:per_cpu])` to `cpu_snapshot_t`
-  @spec to_cpu_snapshot(any()) :: CpuInfo.cpu_snapshot_t()
-  defp to_cpu_snapshot(cpu_util_result) do
+  defp cpu_snapshot(:Windows) do
+    {cpu_util_result, 0} = System.cmd("wmic", ["cpu", "get", "loadpercentage"])
```
Comment (DominikWolek) [inline lib/beamchmark/suite/cpu/cpu_task.ex]: This is weird - why not just call `cpu_snapshot(Utils.os())` in `run_poll` and get rid of this zero-arity `cpu_snapshot`?

File: lib/beamchmark/utils.ex [hunk 2]
```diff
@@ -0,0 +1,17 @@
+defmodule Beamchmark.Utils do
+  @moduledoc """
+  The module defines utility functions for Beamchmark.
+  """
+
+  @spec os :: :FreeBSD | :Linux | :Windows | :macOS
+  def os() do
+    {_family, name} = :os.type()
```
Comment (DominikWolek) [inline lib/beamchmark/utils.ex]: I'd say this name is not descriptory enough. Maybe `os_name`? What do you think?

File: lib/beamchmark/suite/system_info.ex [hunk 3]
```diff
@@ -21,20 +23,9 @@ defmodule Beamchmark.Suite.SystemInfo do
       elixir_version: System.version(),
       otp_version: :erlang.system_info(:otp_release) |> List.to_string(),
       nif_version: :erlang.system_info(:nif_version) |> List.to_string(),
-      os: os(),
+      os: Utils.os(),
```
Comment (DominikWolek) [inline lib/beamchmark/suite/system_info.ex]: Again, maybe `os_name`, or something similar?

File: lib/beamchmark/suite/cpu/cpu_task.ex [hunk 4]
```diff
@@ -13,6 +13,7 @@ defmodule Beamchmark.Suite.CPU.CpuTask do
   """
```
Comment (DominikWolek) [inline lib/beamchmark/suite/cpu/cpu_task.ex]: I think there should be an information in README about which systems are supported.

File: lib/beamchmark/formatters/html.ex [hunk 5]
```diff
@@ -69,10 +72,17 @@ defmodule Beamchmark.Formatters.HTML do
...
   defp get_browser() do
-    case :os.type() do
-      {:unix, :darwin} -> "open"
-      {:unix, _} -> "xdg-open"
-      {:win32, _} -> "explorer"
+    case Utils.os() do
+      :macOS -> "open"
+      :Windows -> "explorer"
+      :Linux -> "xdg-open"
+    end
```
Comment (DominikWolek) [inline lib/beamchmark/formatters/html.ex]: There's no case clause for FreeBSD - this case will throw an error. 

AFAIK `xdg-open` is available on the FreeBSD, and probably whole beamchmark will work on it. But, if you don't want to support it, maybe you should throw the custom error with explanation, that this system is not supported.

Also, I think it's a good idea to describe which systems are supported in the README file.

File: lib/beamchmark/suite/cpu/cpu_task.ex [hunk 6]
```diff
@@ -56,12 +57,27 @@ defmodule Beamchmark.Suite.CPU.CpuTask do
...
+    {cpu_util_result, 0} = System.cmd("wmic", ["cpu", "get", "loadpercentage"])
+
+    average_all_cores =
+      cpu_util_result
+      |> String.split("\r\r\n")
+      |> Enum.at(1)
+      |> String.trim()
+      |> Integer.parse()
+      |> elem(0)
+      |> Kernel./(1)
```
Comment (DominikWolek) [inline lib/beamchmark/suite/cpu/cpu_task.ex]: Why not just use `Float.parse/1`?

**Review by DominikWolek — CHANGES_REQUESTED** [review]

File: lib/beamchmark/suite/cpu/cpu_task.ex (hunk 1 above)
Comment (roznawsk) [reply lib/beamchmark/suite/cpu/cpu_task.ex]: Hmm, in the current approach I like the fact that we don't have to bother with the operating system in the `run_poll`, just calling `cpu_snapshot/0` which checks os for us.

File: lib/beamchmark/utils.ex (hunk 2 above)
Comment (roznawsk) [reply lib/beamchmark/utils.ex]: I will go with `get_os_name`

**Review by roznawsk — COMMENTED** [review]
Due to windows taking around a second to get cpu utilization (averaged from that second), I implemented `run_poll` separately for Windows by making async queries for cpu utilization.

File: README.md [hunk 7]
```diff
@@ -212,7 +213,7 @@ Currently, you can output Beamchmark reports in the following ways:
 
 * Custom formatters
 
-  You can implement your custom formatters by overriding `Beamchmark.Formatter` behaviour.
+  You can implement your custom formatters by overriding `Beamchmark.Formatter` behavior.
```
Comment (DominikWolek) [inline README.md]: As far as I remember we stick to the British English spelling in the documentation, let's keep the "behaviour".

File: lib/beamchmark/suite/cpu/cpu_task.ex [hunk 8]
```diff
@@ -54,14 +72,53 @@ defmodule Beamchmark.Suite.CPU.CpuTask do
...
+    case snapshots_no do
+      0 ->
+        cpu_snapshots
+
+      _snapshots_no ->
+        cpu_snapshots =
+          receive do
+            {:cpu_snapshot, snapshot} ->
+              IO.puts("received snapshot")
+              [snapshot | cpu_snapshots]
```
Comment (DominikWolek) [inline lib/beamchmark/suite/cpu/cpu_task.ex]: I think this is a debug message left out

File: lib/beamchmark/suite/cpu/cpu_task.ex [hunk 9]
```diff
@@ -54,14 +72,53 @@ defmodule Beamchmark.Suite.CPU.CpuTask do
...
+          end
+
+        receive_snapshots(snapshots_no - 1, cpu_snapshots)
+    end
+  end
+
+  @spec cpu_snapshot(:Windows, pid()) :: nil
+  def(cpu_snapshot(:Windows, pid)) do
```
Comment (DominikWolek) [inline lib/beamchmark/suite/cpu/cpu_task.ex]:
```suggestion
  defp cpu_snapshot(:Windows, pid) do
```

Comment (DominikWolek) [inline lib/beamchmark/suite/cpu/cpu_task.ex]: This is still little weird for me. Maybe just change is to something like `windows_cpu_snapshot(pid)`? First argument is ignored anyway.

**Review by DominikWolek — APPROVED** [review]
image

**Pull request merged**